Don't display the editor form if there's no editor field

kienstra opened the pull request:
<!--- Please summarize this PR in the title above -->

#### Changes
* If there are only Inspector fields, this doesn't display the editor form:
![the-ins](https://user-images.githubusercontent.com/4063887/108272677-323e8f00-7138-11eb-8e06-da7d863f83af.gif)
* Branched off https://github.com/studiopress/genesis-custom-blocks/pull/63. That accounts for most of this PR's diff. Here's the [diff](https://github.com/studiopress/genesis-custom-blocks/compare/6a2f8f692c3d34d61c7ac0271dd05213985c6a04...834cbd4c3e4df4cf4cbc40cf8b35fc8e986f0a29) for this feature only.

#### Testing instructions
1. `/wp-admin` > Custom Blocks > Create New
2. Add a field with the location of 'Inspector'
3. Publish the block
4. Create a new post
5. Add the block
6. Expected: the `<ServerSideRender>` displays, as shows in the `.gif` above. The block form never displays.

kienstra committed:
Begin 'Editor Preview', cherry-picking b48e9c6f

Add a tab
for an 'Editor Preview' to the GCB editor.

kienstra committed:
Use wp-env instead of env, so CircleCI can pass

kienstra committed:
On changing to 'Editor Mode', switch to 'Editor' location if needed

The Inspector fields won't display,
so might as well switch to the
'Editor' location.

kienstra committed:
Use editor styles in the GCB editor

For now, just adds a selector
to the stylesheet.

kienstra committed:
Add initial version of 'Front-end Preview'

It only saves the entered values
in the state, and they don't persist across page loads.
It sounds like eventually they should persist.

kienstra committed:
Only show location buttons for 'Builder' mode

kienstra committed:
Display Image field by removing <MediaUploadCheck>

That requires <BlockEditorProvider>,
which means that
getEditedPostContent()
will return [] instead of the
needed JSON string.
That context parses the content into blocks.
But this editor doesn't use blocks.
Still, maybe it's not a good idea
to remove the check.

kienstra committed:
Remove the focus outline, there's still a font-semibold when selected

That's applied via JavaScript.

kienstra committed:
Also show Inspector fields in 'Front-end Preview'

Before, it only showed the fields
for the currentLocation.

kienstra committed:
Don't show the Front-end Preview if it's not published

It won't be available in the block editor
if it's not published,
because it won't be registered
via register_block_type().

kienstra committed:
Simplify the error red border in the Number field

Instead of hacking the activeElement,
add a class to the component.
Todo: apply this to email and other fields.
I could only get the <input type=number> to be
an error on Safari, not Chrome
or Firefox.

kienstra committed:
Simplify the error state for the Email and Url controls

Instead of hacking currentTarget,
simply add a class to the componetn.

kienstra committed:
Remove options prop from <CheckboxControl>

It looks like that doesn't have
an options prop.

kienstra committed:
Remove Jest tests for the error class

This is just testing that the mock functions
are called and add the class.
It's an implementation detail.

kienstra committed:
Get the default export as classNames

That is the name of the package
and what the docs have:
https://github.com/JedWatson/classnames/tree/b166fd132d4a46b6ed93b3264ca638c7f14d4620

kienstra committed:
Show the <ServerSideRender> even if the post isn't published

In the 'Front-end Preview',
the block post might not be published.
But it should still show a preview
using the <ServerSideRender>

kienstra committed:
Simplify the Color control

Before, the Popover seemed to move
in the Popover.Slot.

kienstra committed:
Begin refactoing Image control

This didn't work in the
'Editor Preview',
maybe it needs some modernization.
It was partially forked from
the Image block, which
has since changed a lot.

kienstra committed:
Get the image url on uploading one

Before, this wasn't getting the url.
So look for .url instead
of .source_url

kienstra committed:
Move components/image.js to controls/image.js

components/image.js was just a wrapper
that used <Image>,
so move that wrapper into the component.

kienstra committed:
Move the context logic into a custom hook useImage()

To keep the logic
as separate as possible
from the presentation layer (return statement).

kienstra committed:
Prevent testing warning of unresolved promise rejection

The Jest test for image passed,
but had a warning about
window.fetch rejecting without
handling the failure.

kienstra committed:
Simplify the color component

This was 2 components,
but it can be 1.

kienstra committed:
Merge branch 'develop' into add/editor-preview

kienstra committed:
Make the Editor Preview edits persists across page loads

kienstra committed:
Create ChangeBlock typedef so it's only in one place

Before, this function
had 2 type definitions.

kienstra committed:
Load the user's block CSS file, if one exists

As Rob mentioned,
this is probably an important feature.

kienstra committed:
Only display the block form when there's an edit field

As Rob mentioned,
sometimes there are blocks
with only Inspector fields.
Then, no need to display the editor form,
as there's nothing to show there.

kienstra committed:
Fix a (non-flaky) e2e test failure

The problem was in the logic.
A field can either not have a location,
or have a location of 'editor'.
Both mean they will display in the editor.

kienstra committed:
Change hasEditorField to a boolean, not a function

It doesn't need to be a function,
and it'll remove 2 lines of code.

lukecarbis commented:
Nice! I won't have time to review this until next week unfortunately.

kienstra commented:
Hi @lukecarbis,
All good! 

Do you agree with the idea of this PR (without deep-diving on the code, of course)?

lukecarbis commented:
Yes!

lukecarbis commented on js/src/block-editor/controls/text.js:
```diff
@@ -1,3 +1,8 @@
+/**
+ * External dependencies
+ */
+import * as React from 'react';
```
Hey! How is this useful? Noticed you included it in all the controls.

kienstra replied:
Yeah, it's a little dirty having this in every React file.

But it lets VS Code do type checking. This has revealed some incorrect props, like [here](https://github.com/studiopress/genesis-custom-blocks/pull/63/files#diff-3123e4746cd8f36584d725dcc80b1ef6041d37a82735bc5ef6a1b71bbb50899aL11)

lukecarbis commented on js/src/block-editor/controls/image.js:
```diff
@@ -1,11 +1,117 @@
+/**
```
Why is this moved from components to controls?

kienstra replied:
Before, this file was just a wrapper for `controls/components/image.js`:

https://github.com/studiopress/genesis-custom-blocks/blob/b1062267f4166726ce80581ab26fff5d0aaaf08e/js/src/block-editor/controls/image.js#L1-L12

So I though it'd be good to just move that file into this.

kienstra replied:
Normally, this refactoring of `<Image>` would be way out of scope. 

But I noticed that the Image control didn't display the image in the 'Editor Preview', so figured I'd work on it.

lukecarbis commented on js/src/block-editor/controls/color.js:
```diff
@@ -76,11 +33,33 @@ const GcbColorControl = ( props ) => {
+						onClick={ ( event ) => event.stopPropagation() }
+						onClose={ () => setIsVisible( false ) }
+					>
+						<ColorPicker
```
Maybe not for this PR, but could we use the default theme colours here by pulling them via:
`wp.data.select( 'core/block-editor' ).getSettings().colors`?

kienstra replied:
Oh interesting, good point

lukecarbis replied:
Feel free to resolve, as I guess it's out of scope.

kienstra replied:
Sure, thanks

lukecarbis commented on js/src/block-editor/hooks/useImage.js:
```diff
@@ -0,0 +1,119 @@
+	};
+};
+
+export default useImage;
```
What does this do?

kienstra replied:
It's handles the business logic for the Image control that used to be in [higher-order-components](https://github.com/studiopress/genesis-custom-blocks/blob/b1062267f4166726ce80581ab26fff5d0aaaf08e/js/src/block-editor/components/image.js#L14-L80 like `withSelect()` and `withState()`.

kienstra replied:
Normally, this refactoring of `<Image>` would be way out of scope. 

But I noticed that the Image control didn't display the image in the 'Editor Preview', so figured I'd work on it. I didn't notice any issue in the block editor.

lukecarbis commented on js/src/edit-block/components/field.js:
```diff
@@ -2,7 +2,7 @@
  */
 import * as React from 'react';
-import className from 'classnames';
+import classNames from 'classnames';
```
Good pickup!

kienstra replied:
Haha, thanks!

lukecarbis commented on js/src/common/constants/index.js:
```diff
@@ -0,0 +1,18 @@
+
+export const ALTERNATE_LOCATION = 'inspector';
+export const DEFAULT_LOCATION = 'editor';
+export const LOCATIONS = [ DEFAULT_LOCATION, 'inspector' ];
```
I'm not sure I like this. 🤔 

What's the purpose of abstracting "locations"? Using "Alternate" implies there could only be two locations, which is true… so why not just call them Editor and Inspector?

kienstra replied:
Yeah, that's a good point.

There are a few places where it's good to know that it's a default location, like line 215:

https://github.com/studiopress/genesis-custom-blocks/blob/b1062267f4166726ce80581ab26fff5d0aaaf08e/js/src/edit-block/hooks/useField.js#L204-L217

https://github.com/studiopress/genesis-custom-blocks/blob/b1062267f4166726ce80581ab26fff5d0aaaf08e/js/src/edit-block/components/editor.js#L84

But there might be more places where `DEFAULT_LOCATION` is more confusing than `EDITOR_LOCATION`

I could change the constant names if you'd like. As long as the value of the constant stays the same, it'll work the same whether it's named `EDITOR_LOCATION` or `DEFAULT_LOCATION`

lukecarbis replied:
Of course. Totally your call on this one. Feel free to resolve.

kienstra replied:
If it's alright, https://github.com/studiopress/genesis-custom-blocks/pull/64/commits/60337ef885d98e5fe4a121295db889a3f79c165b renames `ALTERNATE_LOCATION` to `INSPECTOR_LOCATION`, and replaces 2 uses of `DEFAULT_LOCATION` with `EDITOR_LOCATION`. 

Good call on that, default and alternate don't really describe them.

lukecarbis reviewed (approved):
Hey @kienstra! What a great improvement! There are obviously lots of changes here, so I've asked a few clarifying questions in my review, but your work is infuriatingly immaculate as always.

Obviously I've tested the patch as well, and it all works perfectly on my end. Well done! 🎉

kienstra commented:
> There are obviously lots of changes here, so I've asked a few clarifying questions in my review, but your work is infuriatingly immaculate as always.

Haha, thanks so much!

lukecarbis commented:
Feel free to merge when ready.

kienstra commented:
Thanks, Luke!!! 🥳 🎂

kienstra committed:
Replace most DEFAULT_LOCATION uses to EDITOR_LOCATION

Also, remove ALTERNATE_LOCATION
that was confusing, so replace it with
INSPECTOR_LOCATION
